src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,426 @@
+    fn from(ecc_config: &EccConfig) -> Self {
+        let config = Self {
+            q_init_z: ecc_config.q_init_z,
+            q_mul_z: ecc_config.q_mul_z,
+            perm: ecc_config.perm.clone(),
+            add_config: ecc_config.into(),
+            hi_config: ecc_config.into(),
+            lo_config: ecc_config.into(),
+            complete_config: ecc_config.into(),
+            overflow_config: ecc_config.into(),
+        };
+
+        assert_eq!(
+            config.hi_config.x_p, config.lo_config.x_p,
+            "x_p is shared across hi and lo halves."
+        );
+        assert_eq!(
+            config.hi_config.y_p, config.lo_config.y_p,
+            "y_p is shared across hi and lo halves."
+        );
+
+        let add_config_advices = config.add_config.advice_columns();
+        assert!(
+            !add_config_advices.contains(&config.hi_config.z),
+            "hi_config z cannot overlap with complete addition columns."
+        );
+
+        config
+    }
+}
+
+impl Config {
+    pub(super) fn create_gate(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        // Gate used to check that the running sum for scalar decomposition is initialized to zero.
+        meta.create_gate("Initialize running sum for variable-base mul", |meta| {
+            let q_init_z = meta.query_selector(self.q_init_z);
+            let z = meta.query_advice(self.hi_config.z, Rotation::cur());
+
+            vec![q_init_z * z]
+        });
+
+        self.hi_config.create_gate(meta);
+        self.lo_config.create_gate(meta);
+        self.complete_config.create_gate(meta);
+        self.overflow_config.create_gate(meta);
+    }
+
+    pub(super) fn assign(
+        &self,
+        mut layouter: impl Layouter<pallas::Base>,
+        alpha: EccScalarVar,
+        base: &EccPoint,
+    ) -> Result<EccPoint, Error> {
+        let (result, zs): (EccPoint, Vec<Z<pallas::Base>>) = layouter.assign_region(
+            || "variable-base scalar mul",
+            |mut region| {
+                let offset = 0;
+                // Decompose `k = alpha + t_q` bitwise (big-endian bit order).
+                let bits = decompose_for_scalar_mul(alpha.value());
+
+                // Define ranges for each part of the algorithm.
+                let bits_incomplete_hi = &bits[INCOMPLETE_HI_RANGE];
+                let bits_incomplete_lo = &bits[INCOMPLETE_LO_RANGE];
+                let lsb = bits[pallas::Scalar::NUM_BITS as usize - 1];
+
+                // Initialize the accumulator `acc = [2]base`
+                let acc = self
+                    .add_config
+                    .assign_region(&base, &base, offset, &mut region)?;
+
+                // Increase the offset by 1 after complete addition.
+                let offset = offset + 1;
+
+                // Initialize the running sum for scalar decomposition to zero
+                let z_init = {
+                    // Constrain the initialization of `z` to equal zero.
```

Comment:
Let's make a TODO to enforce this to equal zero via public inputs or something (using permutation argument).

Reply:
```suggestion
                    // Constrain the initialization of `z` to equal zero.
                    // TODO: use an equality constraint instead.
```